src/drivers/rgbled_ncp5623c/rgbled_ncp5623c.cpp

```diff
@@ -0,0 +1,423 @@
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+/**
+ * @file rgbled.cpp
+ *
+ * Driver for the onboard RGB LED controller (NCP5623C) connected via I2C.
+ */
+
+#include <px4_config.h>
+#include <px4_getopt.h>
+
+#include <drivers/device/i2c.h>
+
+#include <sys/types.h>
+#include <stdint.h>
+#include <string.h>
+#include <stdlib.h>
+#include <stdbool.h>
+#include <fcntl.h>
+#include <unistd.h>
+#include <stdio.h>
+#include <ctype.h>
+
+#include <px4_workqueue.h>
+
+#include <perf/perf_counter.h>
+#include <systemlib/err.h>
+
+#include <board_config.h>
+
+#include <drivers/drv_led.h>
+#include <lib/led/led.h>
+
+#include "uORB/topics/parameter_update.h"
+
+#define ADDR				 0x39	    /**< I2C adress of NCP5623C */
+
+#define NCP5623_LED_CURRENT  0x20    // Current register
+#define NCP5623_LED_PWM0     0x40    // pwm0 register
+#define NCP5623_LED_PWM1     0x60    // pwm1 register
+#define NCP5623_LED_PWM2     0x80    // pwm2 register
+
+#define NCP5623_LED_BRIGHT  0x1f    // full brightness
+#define NCP5623_LED_OFF     0x00    // off
+
+
+class RGBLED_NPC5623C : public device::I2C
+{
+public:
+	RGBLED_NPC5623C(int bus, int rgbled);
+	virtual ~RGBLED_NPC5623C();
+
+
+	virtual int		init();
+	virtual int		probe();
+private:
+	work_s			_work;
+
+	float			_brightness;
+	float			_max_brightness;
+
+	uint8_t			_r;
+	uint8_t			_g;
+	uint8_t			_b;
+	volatile bool		_running;
+	volatile bool		_should_run;
+	bool			_leds_enabled;
+	int			_param_sub;
+
+	LedController		_led_controller;
+
+	static void		led_trampoline(void *arg);
+	void			led();
+
+	int			send_led_rgb();
+	void		update_params();
+
+	int write(uint8_t reg, uint8_t data);
+};
+
+/* for now, we only support one RGBLED */
+namespace
+{
+RGBLED_NPC5623C *g_rgbled = nullptr;
+}
+
+void rgbled_ncp5623c_usage();
+
+extern "C" __EXPORT int rgbled_ncp5623c_main(int argc, char *argv[]);
+
+RGBLED_NPC5623C::RGBLED_NPC5623C(int bus, int rgbled) :
+	I2C("rgbled_ncp5623c", RGBLED_NCP5623C_DEVICE_PATH, bus, rgbled
+#ifdef __PX4_NUTTX
+	    , 100000 /* maximum speed supported */
+#endif
+	   ),
+	_work{},
+	_brightness(1.0f),
+	_max_brightness(1.0f),
+	_r(0),
+	_g(0),
+	_b(0),
+	_running(false),
+	_should_run(true),
+	_leds_enabled(true),
+	_param_sub(-1)
+{
+}
+
+RGBLED_NPC5623C::~RGBLED_NPC5623C()
+{
+	_should_run = false;
+	int counter = 0;
+
+	while (_running && ++counter < 10) {
+		usleep(100000);
+	}
+}
+
+int RGBLED_NPC5623C::write(uint8_t reg, uint8_t data)
+{
+	uint8_t msg[1] = { 0x00 };
+	msg[0] = ((reg & 0xe0) | (data & 0x1f));
+
+	int ret = transfer(&msg[0], 1, nullptr, 0);
+	return ret;
+}
+
+int
+RGBLED_NPC5623C::init()
+{
+	int ret;
+	ret = I2C::init();
+
+	if (ret != OK) {
+		return ret;
+	}
+
+	update_params();
+
+	_running = true;
+	work_queue(LPWORK, &_work, (worker_t)&RGBLED_NPC5623C::led_trampoline, this, 0);
+	return OK;
+}
+
+int
+RGBLED_NPC5623C::probe()
+{
+	_retries = 4;
+
+	return write(NCP5623_LED_CURRENT, 0x00);
```

Comment:
Just a comment:  probe usually tries to read a `whoami` code from a register. The chip does not have one: sigh! 

perhaps NCP5623_LED_OFF should be used here. (we may change this after testing)